pkg/shares/split_compact_shares.go

```diff
@@ -130,14 +132,28 @@ func (css *CompactShareSplitter) stackPending() {
 
 // Export finalizes and returns the underlying compact shares.
 func (css *CompactShareSplitter) Export() NamespacedShares {
+	if css.isEmpty() {
+		return []NamespacedShare{}
+	}
+
+	var bytesOfPadding int
 	// add the pending share to the current shares before returning
-	if len(css.pendingShare.Share) > appconsts.NamespaceSize+appconsts.ShareInfoBytes {
-		css.pendingShare.Share, _ = zeroPadIfNecessary(css.pendingShare.Share, appconsts.ShareSize)
+	if !css.isEmptyPendingShare() {
+		css.pendingShare.Share, bytesOfPadding = zeroPadIfNecessary(css.pendingShare.Share, appconsts.ShareSize)
 		css.shares = append(css.shares, css.pendingShare)
 	}
-	// force the last share to have a reserve byte of zero
+
+	dataLengthVarint := css.dataLengthVarint(bytesOfPadding)
+	css.writeDataLengthVarintToFirstShare(dataLengthVarint)
+	css.forceLastShareReserveByteToZero()
+	return css.shares
+}
+
+// forceLastShareReserveByteToZero overwrites the reserve byte of the last share
+// with zero. See https://github.com/celestiaorg/celestia-app/issues/779
+func (css *CompactShareSplitter) forceLastShareReserveByteToZero() {
```

Comment:
really good thinking abstracting this

Reply:
hopefully we can remove entirely when we tackle https://github.com/celestiaorg/celestia-app/issues/779

Reply:
yeah, come to think of it, we should finish #779 before the next upgrade as well, since it is also consensus breaking